### PR TITLE
New internal Context API

### LINKED ISSUE #4: New Context API
Right now, adding a value to Roact's context API looks like this:

```lua
function MyComponent:init()
	self._context.blah = "foo"
end
```

Since we originally built Roact, the React team designed and implemented an [entirely new context API](https://reactjs.org/docs/context.html) that's a lot nicer. I think we should steal it wholesale, since it's exactly what we recommend people build on top of `_context` anyways:

```lua
local ThemeContext = Roact.createContext("default value")

local function UsingValue()
	return Roact.createElement(ThemeContext.Consumer, {
		render = function(theme)
			return Roact.createElement("TextLabel", {
				Text = theme,
			})
		end,
	})

	-- We could also introduce a helper live we've done for other context-using
	-- libraries in the past:

	return ThemeContext.with(function(theme)
		return Roact.createElement("TextLabel", {
			Text = theme,
		})
	end)
end

local function Root()
	return Roact.createElement(ThemeContext.Provider, {
		value = "overridden value",
	}, {
		Inner = Roact.createElement(UsingValue),
	})
end
```

Some notable changes from the old context API:
* No ability to explicitly access the internal context table, only the keys you've created
* Context values will update when the provider is updated (!!)
* Not possible to pull context values without also subscribing to their updates

I'm on the fence a bit about the `contextType` API that React has, but it would be useful for components that need to use context values outside of `render`.

### PR DESCRIPTION
**Pull request by ZoteTheMighty**
Works towards #4. This is a prerequisite for the new _public_ Context API proposed in #246.

This change introduced two new methods to the `Component` API:

* `Component:__addContext(key, value)` - Adds a new entry to the context table. This will be available to this component and all of its children, and is intended to be used in a Provider component.
* `Component:__getContext(key)` - Retrieves the context value for the given key. This intended to be used by a Consumer component.

This interface allows us to avoid defensive copying, as even internal Component APIs do not get direct access to the context table. The legacy context API is still present and usable, and will not be affected by these changes.

Checklist before submitting:
* ~~Added entry to `CHANGELOG.md`~~ (this should come with the changes in #246)
* [x] Added/updated relevant tests
* [x] Added/updated documentation (documentation here consists of code comments, public docs will be part of #246)

### CODE REVIEW

File: src/Component.spec/context.spec.lua [hunk 1]
```diff
@@ -92,12 +99,85 @@ return function()
...
+		local expectedContext = {
+			parent = "I'm here!",
+		}
+
+		assertDeepEqual(parentNode.context, expectedContext)
+
+		local childNode = oneChild(parentNode.children)
+
+		-- Parent and child should have the same context table
+		expect(parentNode.context, childNode.context)
```
Comment (LPGhatguy) [inline src/Component.spec/context.spec.lua]: Do you mean to use `assert(a == b)` here?

File: src/createReconciler.lua [hunk 2]
```diff
@@ -247,10 +258,14 @@ local function createReconciler(renderer)
...
 		Constructs a new virtual node but not does mount it.
 	]]
-	local function createVirtualNode(element, hostParent, hostKey, context)
+	local function createVirtualNode(element, hostParent, hostKey, context, legacyContext)
 		if config.internalTypeChecks then
 			internalAssert(renderer.isHostObject(hostParent) or hostParent == nil, "Expected arg #2 to be a host object")
 			internalAssert(typeof(context) == "table" or context == nil, "Expected arg #4 to be of type table or nil")
+			internalAssert(
+				typeof(legacyContext) == "table" or legacyContext == nil,
+				"Expected arg #4 to be of type table or nil"
```
Comment (LPGhatguy) [inline src/createReconciler.lua]:
```suggestion
				"Expected arg #5 to be of type table or nil"
```

Typo, this is arg 5.

File: src/createReconciler.lua [hunk 3]
```diff
@@ -304,10 +330,13 @@ local function createReconciler(renderer)
...
 		the tree.
 	]]
-	function mountVirtualNode(element, hostParent, hostKey, context)
+	function mountVirtualNode(element, hostParent, hostKey, context, legacyContext)
 		if config.internalTypeChecks then
 			internalAssert(renderer.isHostObject(hostParent) or hostParent == nil, "Expected arg #2 to be a host object")
-			internalAssert(typeof(context) == "table" or context == nil, "Expected arg #4 to be of type table or nil")
+			internalAssert(
+				typeof(legacyContext) == "table" or legacyContext == nil,
+				"Expected arg #4 to be of type table or nil"
```
Comment (LPGhatguy) [inline src/createReconciler.lua]:
```suggestion
				"Expected arg #5 to be of type table or nil"
```

Same copy-paste typo here.

File: src/Component.spec/context.spec.lua (hunk 1 above)
Comment (ZoteTheMighty) [reply src/Component.spec/context.spec.lua]: I sure did! Fortunately, it still passes.

**Pull request merged**